crates/common/tedge_utils/src/inotify.rs

```diff
@@ -0,0 +1,222 @@
+pub enum InotifyStreamError {
+    #[error(transparent)]
+    FromIOError(#[from] std::io::Error),
+
+    #[error("Error starting inotify service.")]
+    InotifyInitError,
+
+    #[error("Error creating event stream")]
+    ErrorCreatingStream,
+
+    // FIXME: how to show which mask is unsupported when inotify::WatchMask
+    // does not impl Display?
+    #[error("Unsupported mask.")]
+    UnsupportedWatchMask,
+}
+
+impl From<&InotifyStreamWatchMask> for EventMask {
+    fn from(val: &InotifyStreamWatchMask) -> Self {
+        match val {
+            InotifyStreamWatchMask::CloseWrite => EventMask::CLOSE_WRITE,
+            InotifyStreamWatchMask::Delete => EventMask::DELETE,
+            InotifyStreamWatchMask::Created => EventMask::CREATE,
+        }
+    }
+}
+
+impl From<&InotifyStreamWatchMask> for WatchMask {
+    fn from(val: &InotifyStreamWatchMask) -> Self {
+        match val {
+            InotifyStreamWatchMask::CloseWrite => WatchMask::CLOSE_WRITE,
+            InotifyStreamWatchMask::Delete => WatchMask::DELETE,
+            InotifyStreamWatchMask::Created => WatchMask::CREATE,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum Either {
+    FileWatch(Vec<(PathBuf, InotifyStreamWatchMask)>),
+    DirWatch((PathBuf, Vec<InotifyStreamWatchMask>)),
+}
+
+impl Either {
+    fn is_file_watch(&self) -> bool {
+        match self {
+            Self::FileWatch(_) => true,
+            Self::DirWatch(_) => false,
+        }
+    }
+
+    fn is_dir_watch(&self) -> bool {
+        match self {
+            Self::FileWatch(_) => false,
+            Self::DirWatch(_) => true,
+        }
+    }
+}
+
+struct InotifyStream {
+    buffer: [u8; 1024],
+    inotify: Inotify,
+    watcher: Vec<Arc<Either>>,
+}
+
+impl Default for InotifyStream {
+    fn default() -> Self {
+        let inotify = Inotify::init().unwrap();
+        let buffer = [0; 1024];
+        let watcher = vec![Arc::new(Either::FileWatch(vec![]))];
+        Self {
+            buffer,
+            inotify,
+            watcher,
+        }
+    }
+}
+
+impl InotifyStream {
+    pub fn add_watcher(
+        &mut self,
+        path: &Path,
+        mask: InotifyStreamWatchMask,
+    ) -> Result<(), InotifyStreamError> {
+        let watcher = if path.is_dir() {
+            Arc::new(Either::DirWatch((path.to_path_buf(), vec![mask])))
+        } else {
+            Arc::new(Either::FileWatch(vec![(path.to_path_buf(), mask)]))
+        };
+        let watcher_clone = watcher.clone();
+
+        match &*watcher {
+            Either::FileWatch(files_to_watch) => {
+                for (file_path, mask) in files_to_watch {
+                    let parent_dir = file_path
+                        .parent()
+                        .ok_or_else(|| InotifyStreamError::ErrorCreatingStream)?;
+                    //let _ = self.inotify.add_watch(parent_dir, mask.into())?;
+                    let _ = self.inotify.add_watch(parent_dir, WatchMask::ALL_EVENTS)?;
```

Comment:
Note here, that `WatchMask::ALL_EVENTS` has been added here as a way of not overwriting events when watching the same `parent_dir`. according the inotify API `add_watch`:
> Adds a new watch or updates an existing one for the file referred to by path. Returns a watch descriptor that can be used to refer to this watch later.